adapter/internal/ingress/client_manager.go

```diff
@@ -64,21 +65,29 @@ func NewClientManager(connector ConnectionBuilder, connCount int, ttl time.Durat
 func (c *ClientManager) Next() v2.EgressClient {
 	for {
 		idx := atomic.AddUint64(&c.nextIdx, 1)
-		conn := (*connection)(atomic.LoadPointer(&c.connections[idx%uint64(len(c.connections))]))
-		if conn.client != nil {
+
+		c.mu.RLock()
+		actualIdx := int(idx % uint64(len(c.connections)))
+		conn := c.connections[actualIdx]
+		c.mu.RUnlock()
+
+		if conn != nil && conn.client != nil {
 			return conn.client
 		}
 
+		c.openNewConnection(actualIdx)
```

Comment:
Another concern I had with this is if you have multiple goroutines calling `Next` and they happen to land on the same index and see that it is nil, each goroutine will attempt to create a connection and last write wins. This is at best wasteful. It also means `Close` wouldn't get called for the connections that didn't win. Is that ok?